<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>ChIP-seq coverage analysis</title>
  <meta name="description" content="UPDATE">

  <link rel="stylesheet" href="/adomingues.github.io//css/main.css">
  <link rel="canonical" href="http://localhost:4000/adomingues.github.io//2013/06/02/chip-seq-coverage-analysis.html">
  <link rel="alternate" type="application/rss+xml" title="Moving To The Dark Side" href="http://localhost:4000/adomingues.github.io//feed.xml" />
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/adomingues.github.io//">Moving To The Dark Side</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        
          
          <a class="page-link" href="/adomingues.github.io//about/">About</a>
          
        
          
        
          
        
          
        
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <header class="post-header">
    <h1 class="post-title">ChIP-seq coverage analysis</h1>
    <p class="post-meta">Jun 2, 2013</p>
  </header>

  <article class="post-content">
    <p>UPDATE</p>

<p>It has been a long time since I wrote this post, and in between a wonderfull set of tools for coverage calculation and visualization has been published: <a href="https://github.com/fidelram/deepTools/wiki">deepTools</a> does metagene analysis, and much, much more in a better than I could possibly do with my näive scripts.</p>

<p>Try it. It is worth it.</p>

<p>There a few analysis tasks for which one would think a solution, or an easy to follow example, is readily available in the web (it has been a long time since the Web became the Internet). One such example is how to do coverage analysis of ChIP-seq, or any other data for that matter. That is, how are the reads distributed around a region of interest, commonly the TSS?</p>

<p><a href="http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html" title="HTSeq overview">HTSeq</a>, the python library for NGS data analysis offers <a href="http://www-huber.embl.de/users/anders/HTSeq/doc/tss.html#tss" title="HTSeq - TSS analysis">that</a>, but I prefer to do my plots in R and ggplot. I also find <a href="http://www.bioconductor.org/help/course-materials/2009/EMBLJune09/Practicals/TSS/TSS_plot.pdf" title="TSS - An exercise with coverage vectors">bioconductor</a> nice but some of the objects and the packages literature hard to understand for someone looking for imminently practical information.  So I decided to ask around in my institute and create my own script. I like to do scripts that can be easily modified and also that generate plots that can be used with little or now change for presentations.</p>

<p><strong>Create the bed file with TSSs:</strong></p>

<p>Firstly, how to create the BED file with the TSSs using a combination of bioconductor and <a href="http://code.google.com/p/bedtools/">bedtools</a> - it is quite fast and it very-well documented. This only needs to be ran once and there is a myriad of ways to do it. Of course one could also use any bed file with features of interest.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1">######################  </span>
<span class="c1">### Load libraries ###  </span>
<span class="kn">library</span><span class="p">(</span>GenomicRanges<span class="p">)</span>  
<span class="kn">library</span><span class="p">(</span>GenomicFeatures<span class="p">)</span>  
<span class="kn">library</span><span class="p">(</span>plyr<span class="p">)</span>  
<span class="kn">library</span><span class="p">(</span>reshape<span class="p">)</span>  
<span class="kn">library</span><span class="p">(</span>data.table<span class="p">)</span>

<span class="c1">### custom functions ##  </span>
<span class="c1"># string concatenation without space gaps  </span>
concat <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span><span class="kc">...</span><span class="p">)</span> <span class="kp">paste</span><span class="p">(</span><span class="kc">...</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>

<span class="c1">################################  </span>
<span class="c1">## create TSS bed file (UCSC) ##  </span>
<span class="c1"># create a database of transcripts  </span>
hg19RefGenes <span class="o">&lt;-</span> makeTranscriptDbFromUCSC<span class="p">(</span>genome <span class="o">=</span> <span class="s">&quot;hg19&quot;</span><span class="p">,</span> tablename <span class="o">=</span> <span class="s">&quot;refGene&quot;</span><span class="p">)</span> <span class="c1"># run only once  </span>
hg19RefGenes <span class="c1"># summary of the object  </span>
saveDb<span class="p">(</span>hg19RefGenes<span class="p">,</span> file<span class="o">=</span><span class="s">&#39;~/Homo\_sapiens/hg19\_RefSeq\_genes.sqlite&#39;</span><span class="p">)</span>  
<span class="c1"># by saving the database one keeps a record of which version of the genome was used for the analysis. Very important for reproducibility.  </span>
hg19RefGenes <span class="o">&lt;-</span> loadDb<span class="p">(</span>file<span class="o">=</span><span class="s">&quot;~/Homo\_sapiens/hg19\_RefSeq\_genes.sqlite&quot;</span><span class="p">)</span> <span class="c1"># not necessary of useful if the the script need to be ran again later</span>

<span class="c1"># create an object with a list of transcripts grouped by gene  </span>
transcripts <span class="o">&lt;-</span> transcriptsBy<span class="p">(</span>hg19RefGenes<span class="p">,</span> by<span class="o">=</span><span class="s">&quot;gene&quot;</span><span class="p">)</span>

<span class="c1"># convert to data frame:  </span>
gene2tx <span class="o">&lt;-</span> <span class="kp">as.data.frame</span><span class="p">(</span>transcripts<span class="p">)</span>

<span class="c1"># Note that transcription\_start is always smaller than transcription\_end, even when the transcript is on the &quot;−&quot; strand. Hence, we have to use either the start or the end coordinate of the transcript, depending on the strand, to get the actual transcription start sites, i.e., the 5&#39; ends of the transcripts:  </span>
<span class="c1">#http://www.bioconductor.org/help/course-materials/2009/EMBLJune09/Practicals/TSS/TSS\_plot.pdf  </span>
tss <span class="o">&lt;-</span> <span class="kp">ifelse</span><span class="p">(</span>gene2tx<span class="o">$</span>strand <span class="o">==</span> <span class="s">&quot;+&quot;</span><span class="p">,</span> gene2tx<span class="o">$</span>start<span class="p">,</span> gene2tx<span class="o">$</span>end<span class="p">)</span>

tss.bed <span class="o">&lt;-</span> <span class="kt">data.frame</span><span class="p">(</span>seqnames <span class="o">=</span> gene2tx<span class="o">$</span>seqnames<span class="p">,</span> start <span class="o">=</span> tss<span class="p">,</span> end<span class="o">=</span> tss<span class="p">,</span> strand<span class="o">=</span>gene2tx<span class="o">$</span>strand<span class="p">,</span> transcriptID<span class="o">=</span>gene2tx<span class="o">$</span>tx\_name<span class="p">,</span> type<span class="o">=</span><span class="kp">rep</span><span class="p">(</span><span class="s">&quot;tss&quot;</span><span class="p">,</span> <span class="kp">length</span><span class="p">(</span>tss<span class="p">)))</span>

<span class="c1">## just keep proper chromosomes  </span>
tss.bed <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span>tss.bed<span class="p">,</span> str\_detect<span class="p">(</span>seqnames<span class="p">,</span> <span class="s">&quot;^[[:alnum:]]\*$&quot;</span><span class="p">))</span>  
<span class="kp">with</span><span class="p">(</span>tss.bed<span class="p">,</span> <span class="kp">as.data.frame</span><span class="p">(</span><span class="kp">table</span><span class="p">(</span>seqnames<span class="p">)))</span>    <span class="c1"># list the frequency of genes per chromosome</span>

<span class="c1"># save bed file  </span>
write.table<span class="p">(</span>tss.bed<span class="p">,</span> file <span class="o">=</span> <span class="s">&quot;~/Homo\_sapiens/hg19RefSeqTSS.bed&quot;</span><span class="p">,</span> row.names <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  
col.names <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> quote <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\\t&quot;</span><span class="p">)</span>

<span class="c1">## use bedtools to increase the region around the tss  </span>
<span class="kp">system</span><span class="p">(</span><span class="s">&#39;  </span>
<span class="s">chromInfo=~/Homo\_sapiens/UCSC/hg19/Annotation/Genes/ChromInfo.txt</span>

<span class="s">bedtools slop -i ~/Homo\_sapiens/hg19RefSeqTSS.bed -g $chromInfo -b 1000 \&gt; ~/Homo\_sapiens/hg19RefSeqTSS\_1000kb.bed  </span>
<span class="s">&#39;</span><span class="p">)</span></code></pre></div>

<p><strong>Counting the number of reads per nucleotide</strong></p>

<p>Because I run over several bam files and I like to keep the file names tidy I&#39;ve wrapped the bedtools function in a shell function so that the several analysis can ran in parallel to save time. Because this step takes a of time, it sends me an email at the end. This way I can do something else without being constantly looking at it. Hint: running it in  a <a href="http://www.rackaid.com/resources/linux-screen-tutorial-and-how-to/" title="Intro  to screen sessions">screen session</a> also saves a lot of headaches. If you are doing for a single bam file, or are not versed in shell scripting all you need is to run this:</p>

<div class="highlight"><pre><code class="language-bash" data-lang="bash">coverageBed -d -abam bamFile -b geneModel <span class="se">\&gt;</span> name.cov.bed</code></pre></div>

<p>But in all likelihood you will have several bam files:</p>

<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="c">#################################################################################  </span>
<span class="c"># This little shell script calculates coverage per base at any given annotation #</span>

<span class="c"># wrapper shell function that does the coverage calculations and returns the files with nice names  </span>
coverageAtFeatures<span class="o">(){</span>  
<span class="c"># the input is a bam file that comes from sdout (see bellow)  </span>
<span class="nv">bamFile</span><span class="o">=</span><span class="nv">$1</span>  
<span class="nv">name</span><span class="o">=</span><span class="k">$(</span>basename <span class="k">${</span><span class="nv">bamFile</span><span class="p">%%.bam</span><span class="k">})</span> <span class="c"># to keep the directory tidy I rename the results files using the original bam file  </span>
<span class="nv">geneModel</span><span class="o">=</span><span class="nv">$2</span> <span class="c"># also from sdout  </span>
<span class="nv">outDir</span><span class="o">=</span>results/coverage <span class="c"># defines the output directory  </span>
coverageBed -d -abam <span class="nv">$bamFile</span> -b <span class="nv">$geneModel</span> <span class="se">\&gt;</span> <span class="nv">$outDir</span>/<span class="nv">$name</span>.cov.bed <span class="c"># actually performs the coverage calculation.  </span>
<span class="c"># note the -d, very important!</span>

<span class="o">}</span>  
<span class="nb">export</span> -f coverageAtFeatures <span class="c"># this add the function to the shell to be used</span>

<span class="c"># Features:  </span>
<span class="nv">promoters</span><span class="o">=</span>~/Homo<span class="se">\_</span>sapiens/hg19RefSeqTSS<span class="se">\_</span>1000kb.bed  
<span class="c"># bam folder:  </span>
bam<span class="se">\_</span>folder<span class="o">=</span>~/chip<span class="se">\_</span>seq/data/reads/

<span class="c"># now we can run the function defined above:  </span>
<span class="c"># ls will list all the bam files in the folder and pass them to parallel.  </span>
<span class="c">#Each bam file will the first argument of coverageAtFeatures  </span>
<span class="c"># the second argument will be the bed file with the genomic regions of interest.  </span>
<span class="c"># Parallel will run the analysis for 4 bam simultaneously - this # can vary depending on your computer.  </span>
parallel  -i -j <span class="m">4</span> bash -c <span class="s2">&quot;coverageAtFeatures {} $promoters&quot;</span> --  <span class="k">$(</span>ls <span class="nv">$bam</span><span class="se">\_</span>folder/<span class="se">\*</span>bam<span class="k">)</span><span class="s2">&quot;</span>

<span class="s2"># when all is done a message will arrive in your inbox.  </span>
<span class="s2">echo &quot;</span>Subject:Coverage <span class="k">done</span><span class="err">&quot;</span> <span class="p">|</span>  sendmail -v myname@mymail.com <span class="se">\&gt;</span> /dev/null</code></pre></div>

<p>At the end of this we will have a file that contains something like:</p>

<div class="highlight"><pre><code class="language-bash" data-lang="bash">head chip<span class="se">\_</span>experiment.cov.bed  
chr1 <span class="m">19922471</span> <span class="m">19924471</span> + NM<span class="se">\_</span><span class="m">001204088</span> tss <span class="m">1</span> <span class="m">1</span>  
chr1 <span class="m">19922471</span> <span class="m">19924471</span> + NM<span class="se">\_</span><span class="m">001204088</span> tss <span class="m">2</span> <span class="m">1</span>  
chr1 <span class="m">19922471</span> <span class="m">19924471</span> + NM<span class="se">\_</span><span class="m">001204088</span> tss <span class="m">3</span> <span class="m">1</span>  
chr1 <span class="m">19922471</span> <span class="m">19924471</span> + NM<span class="se">\_</span><span class="m">001204088</span> tss <span class="m">4</span> <span class="m">1</span>  
chr1 <span class="m">19922471</span> <span class="m">19924471</span> + NM<span class="se">\_</span><span class="m">001204088</span> tss <span class="m">5</span> <span class="m">1</span>  
chr1 <span class="m">19922471</span> <span class="m">19924471</span> + NM<span class="se">\_</span><span class="m">001204088</span> tss <span class="m">6</span> <span class="m">0</span>  
chr1 <span class="m">19922471</span> <span class="m">19924471</span> + NM<span class="se">\_</span><span class="m">001204088</span> tss <span class="m">7</span> <span class="m">0</span>  
chr1 <span class="m">19922471</span> <span class="m">19924471</span> + NM<span class="se">\_</span><span class="m">001204088</span> tss <span class="m">8</span> <span class="m">0</span>  
chr1 <span class="m">19922471</span> <span class="m">19924471</span> + NM<span class="se">\_</span><span class="m">001204088</span> tss <span class="m">9</span> <span class="m">0</span>  
chr1 <span class="m">19922471</span> <span class="m">19924471</span> + NM<span class="se">\_</span><span class="m">001204088</span> tss <span class="m">10</span> 0</code></pre></div>

<p>Col1-6 are the original entries of the bed file containing the TSSs, followed by the position of the base in the feature (remember this is a bp-resolution coverage) and the number of reads that mapped to the base.</p>

<p>Now we are ready calculate some statistics on own many reads map to each base surrounding the TSSs and plot these. In the next snippet I&#39;ll calculate the sum of reads per base but calculating the average requires a minor change to the script.  Since several of the operations take some time, I use the package doMC save time but this should work without it.</p>

<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1">############################################  </span>
<span class="c1">## function to calculate coverage profile ##</span>

<span class="c1">## list coverage files  </span>
path<span class="o">=</span><span class="s">&quot;./&quot;</span>  
cov.files <span class="o">&lt;-</span> <span class="kp">list.files</span><span class="p">(</span>path<span class="o">=</span>path<span class="p">,</span> pattern<span class="o">=</span><span class="s">&quot;.cov.bed&quot;</span><span class="p">)</span>

<span class="c1">## for multicore usage:  </span>
<span class="kn">library</span><span class="p">(</span>doMC<span class="p">)</span>  
registerDoMC<span class="p">(</span>cores<span class="o">=</span><span class="m">10</span><span class="p">)</span> <span class="c1"># yay for our server</span>

halfWindow<span class="o">=</span><span class="m">1000</span> <span class="c1"># this half the size of the window around the TSSs</span>

loadCovMat <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span><span class="kp">file</span><span class="p">){</span>  
<span class="c1"># read files  </span>
cov.data <span class="o">&lt;-</span> read.table<span class="p">(</span>concat<span class="p">(</span>path<span class="p">,</span> <span class="kp">file</span><span class="p">),</span> header<span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span> sep <span class="o">=</span> <span class="s">&quot;\\t&quot;</span><span class="p">,</span> stringsAsFactors <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

<span class="c1">#     http://stackoverflow.com/questions/1727772/quickly-reading-very-large-tables-as-dataframes-in-r  </span>
<span class="c1"># cov.data &lt;- fread(file)    # not used because strand not recognized  </span>
<span class="kp">names</span><span class="p">(</span>cov.data<span class="p">)</span> <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span> <span class="s">&quot;start&quot;</span><span class="p">,</span> <span class="s">&quot;end&quot;</span><span class="p">,</span> <span class="s">&quot;strand&quot;</span><span class="p">,</span> <span class="s">&quot;ID&quot;</span><span class="p">,</span> <span class="s">&quot;type&quot;</span><span class="p">,</span> <span class="s">&quot;position&quot;</span><span class="p">,</span> <span class="s">&quot;count&quot;</span><span class="p">)</span>

<span class="c1"># create a new ID to avoid repeated gene names - common issue with UCSC annotations  </span>
cov.data.n <span class="o">&lt;-</span> <span class="kp">transform</span><span class="p">(</span>cov.data<span class="p">,</span> newID <span class="o">=</span> <span class="kp">paste</span><span class="p">(</span>ID<span class="p">,</span> start<span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\_&quot;</span><span class="p">))</span>  
<span class="c1"># add base position with reference to the TSS.  </span>
<span class="c1"># very important: not the strand operation.  </span>
cov.algned <span class="o">&lt;-</span> mutate<span class="p">(</span>cov.data.n<span class="p">,</span> position\_aligned<span class="o">=</span>position<span class="o">-</span>halfWindow<span class="p">,</span>  
position\_str\_cor<span class="o">=</span><span class="kp">ifelse</span><span class="p">(</span>strand<span class="o">==</span><span class="s">&quot;+&quot;</span><span class="p">,</span> position\_aligned<span class="p">,</span> <span class="m">-1</span>\<span class="o">*</span>position\_aligned<span class="p">))</span>

<span class="c1"># for the next operation R needs a lot of memory so I&#39;ll remove objs from memory  </span>
<span class="kp">rm</span><span class="p">(</span><span class="s">&quot;cov.data&quot;</span><span class="p">,</span> <span class="s">&quot;cov.data.n&quot;</span><span class="p">,</span> <span class="s">&quot;cov.data.t&quot;</span><span class="p">)</span>

<span class="c1"># counts the number of reads per position - the final result!  </span>
res <span class="o">&lt;-</span> ddply<span class="p">(</span>cov.algned<span class="p">,</span> <span class="m">.</span><span class="p">(</span>position\_str\_cor<span class="p">),</span> summarize<span class="p">,</span> sum<span class="o">=</span><span class="kp">sum</span><span class="p">(</span>count<span class="p">),</span> <span class="m">.</span>parallel<span class="o">=</span><span class="bp">T</span><span class="p">)</span>

<span class="c1">## get condition name  </span>
<span class="c1"># important to have column with the name of the experiment when running multiple analysis  </span>
sample <span class="o">&lt;-</span> <span class="kp">strsplit</span><span class="p">(</span><span class="kp">file</span><span class="p">,</span> <span class="s">&quot;\\\\.&quot;</span><span class="p">)[[</span><span class="m">1</span><span class="p">]][</span><span class="m">1</span><span class="p">]</span>  
res<span class="o">$</span>condition <span class="o">&lt;-</span> <span class="kp">rep</span><span class="p">(</span><span class="kp">sample</span><span class="p">,</span> <span class="kp">nrow</span><span class="p">(</span>res<span class="p">))</span>

<span class="kr">return</span><span class="p">(</span>res<span class="p">)</span>  
<span class="p">}</span>

<span class="c1">## Run the function over all coverage files  </span>
<span class="c1"># the output is a list of data frames  </span>
cov.l <span class="o">&lt;-</span> <span class="kp">lapply</span><span class="p">(</span>cov.files<span class="p">,</span> loadCovMat<span class="p">)</span>

<span class="c1">## see the data  </span>
<span class="kp">lapply</span><span class="p">(</span>cov.l<span class="p">,</span> <span class="kp">head</span><span class="p">)</span>

<span class="c1">## save the coverage data  </span>
<span class="kp">lapply</span><span class="p">(</span>cov.l<span class="p">,</span> <span class="kr">function</span><span class="p">(</span>x<span class="p">){</span>  
write.table<span class="p">(</span>x<span class="p">,</span>  
concat<span class="p">(</span><span class="kp">unique</span><span class="p">(</span>x<span class="o">$</span>condition<span class="p">),</span> <span class="s">&quot;.sum.cov&quot;</span><span class="p">),</span>  
row.names <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>  
quote <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\\t&quot;</span><span class="p">)})</span>

<span class="c1"># join the data frame in he list in one big data frame for plotting  </span>
cov <span class="o">&lt;-</span> <span class="kp">do.call</span><span class="p">(</span><span class="s">&quot;rbind&quot;</span><span class="p">,</span> cov.l<span class="p">)</span>

<span class="c1">## 2 different plots:  </span>
p.cov.all <span class="o">&lt;-</span> ggplot<span class="p">(</span>data<span class="o">=</span>cov<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>position\_str\_cor<span class="p">,</span> y<span class="o">=</span><span class="kp">sum</span><span class="p">,</span> color<span class="o">=</span>condition<span class="p">))</span> <span class="o">+</span> geom\_line<span class="p">()</span>  
ggsave<span class="p">(</span>filename <span class="o">=</span> <span class="s">&quot;./plots/coverage\_tss\_all.pdf&quot;</span><span class="p">,</span> p.cov.all<span class="p">)</span>

p.cov.wrap <span class="o">&lt;-</span> p.cov.all <span class="o">+</span> facet\_wrap<span class="p">(</span><span class="o">~</span>condition<span class="p">)</span>  
ggsave<span class="p">(</span>filename <span class="o">=</span> <span class="s">&quot;./plots/coverage\_tss\_wrap.pdf&quot;</span><span class="p">,</span> p.cov.wrap<span class="p">)</span></code></pre></div>

<p>If all went well this is the final result:<br>
<a href="http://movingtothedarkside.files.wordpress.com/2013/06/coverage_tss_all_mod.pdf">Plot 1</a></p>

<p><a href="http://movingtothedarkside.files.wordpress.com/2013/06/coverage_tss_wrap_mod.pdf">Plot2</a></p>

<p>Al there is lacking now is putting all of this in a nice wrapper, which could be easily done, but I preferred to show the different stages of the process separated because it is easier to explain. I hope it helps.</p>

  </article>

</div>

<div id="disqus_thread"></div>
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES * * */
    var disqus_shortname = 'adomingues-blog-io';

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">Moving To The Dark Side</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col  footer-col-1">
        <ul class="contact-list">
          <li>Moving To The Dark Side</li>
          <li><a href="mailto:amjdomingues@gmail.com">amjdomingues@gmail.com</a></li>
        </ul>
      </div>

      <div class="footer-col  footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/adomingues">
              <span class="icon  icon--github">
                <svg viewBox="0 0 16 16">
                  <path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/>
                </svg>
              </span>

              <span class="username">adomingues</span>
            </a>
          </li>
          

          
        </ul>
      </div>

      <div class="footer-col  footer-col-3">
        <p class="text"></p>
      </div>
    </div>

  </div>

</footer>


  </body>

  <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-60277379-1', 'auto');
  ga('send', 'pageview');

</script>


</html>
