<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>BioPython musings</title>
  <meta name="description" content="On my quest to improve my scripting skils I have just recently taken the Introduction to BioPython course at the VIB in Leuven.">

  <link rel="stylesheet" href="/adomingues.github.io//css/main.css">
  <link rel="canonical" href="http://localhost:4000/adomingues.github.io//2013/01/04/biopython-musings.html">
  <link rel="alternate" type="application/rss+xml" title="Moving To The Dark Side" href="http://localhost:4000/adomingues.github.io//feed.xml" />
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/adomingues.github.io//">Moving To The Dark Side</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        
          
          <a class="page-link" href="/adomingues.github.io//about/">About</a>
          
        
          
        
          
        
          
        
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <header class="post-header">
    <h1 class="post-title">BioPython musings</h1>
    <p class="post-meta">Jan 4, 2013</p>
  </header>

  <article class="post-content">
    <p>On my quest to improve my scripting skils I have just recently taken the Introduction to <a href="http://www.bits.vib.be/index.php/training/92-python" title="BioPython course">BioPython</a> course at the VIB in Leuven.</p>

<p>The course</p>

<p>It was all very well organized and the Instructor, <a href="http://academis.sites.djangoeurope.com/" title="Instructor">Kristian Rother </a>, was well prepared and engaging. Perhaps he was not expecting so many of wanting to learn more about handing NGS data but when he saw the interest, wrote overnight a custom script to parse and extract information SAM files. Not only it is useful, it also serves a practical example of several Python functions:</p>

<div class="highlight"><pre><code class="language-python" data-lang="python"><span class="c">#!/usr/bin/env python</span>

<span class="c"># TASK: find out how many sequences there are in the filtered dataset  </span>
<span class="c">#       (the one printed at the bottom)</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;example.sam&#39;</span><span class="p">):</span>  
<span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;$&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;@&#39;</span><span class="p">):</span>  
<span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>  
<span class="n">record</span> <span class="o">=</span> <span class="p">{</span>  
<span class="s">&#39;QNAME&#39;</span> <span class="p">:</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>  
<span class="s">&#39;FLAG&#39;</span>  <span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>  
<span class="s">&#39;RNAME&#39;</span> <span class="p">:</span> <span class="n">line</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>  
<span class="s">&#39;POS&#39;</span>   <span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span>  
<span class="s">&#39;MAPQ&#39;</span>  <span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">4</span><span class="p">]),</span>  
<span class="s">&#39;CIGAR&#39;</span> <span class="p">:</span> <span class="n">line</span><span class="p">[</span><span class="mi">5</span><span class="p">],</span>  
<span class="s">&#39;RNEXT&#39;</span> <span class="p">:</span> <span class="n">line</span><span class="p">[</span><span class="mi">6</span><span class="p">],</span>  
<span class="s">&#39;PNEXT&#39;</span> <span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">7</span><span class="p">]),</span>  
<span class="s">&#39;TLEN&#39;</span>  <span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">8</span><span class="p">]),</span>  
<span class="s">&#39;SEQ&#39;</span>   <span class="p">:</span> <span class="n">line</span><span class="p">[</span><span class="mi">9</span><span class="p">],</span>  
<span class="s">&#39;QUAL&#39;</span>  <span class="p">:</span> <span class="n">line</span><span class="p">[</span><span class="mi">10</span><span class="p">],</span>  
<span class="s">&#39;optional&#39;</span> <span class="p">:</span> <span class="p">[]</span>  
<span class="p">}</span>  
<span class="k">for</span> <span class="n">optional</span> <span class="ow">in</span> <span class="n">line</span><span class="p">[</span><span class="mi">11</span><span class="p">:]:</span>  
<span class="n">record</span><span class="p">[</span><span class="s">&#39;optional&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">optional</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">))</span>  
<span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

<span class="c"># filter all records between position 109200 and 110000  </span>
<span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>  
<span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>  
<span class="k">if</span> <span class="n">rec</span><span class="p">[</span><span class="s">&#39;POS&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">109200</span> <span class="ow">and</span> <span class="n">rec</span><span class="p">[</span><span class="s">&#39;POS&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="p">[</span><span class="s">&#39;SEQ&#39;</span><span class="p">])</span> <span class="o">&lt;=</span> <span class="mi">110000</span><span class="p">:</span>  
<span class="k">print</span> <span class="n">rec</span><span class="p">[</span><span class="s">&#39;POS&#39;</span><span class="p">],</span> <span class="n">rec</span><span class="p">[</span><span class="s">&#39;SEQ&#39;</span><span class="p">]</span>  
<span class="n">count</span> <span class="o">=</span> <span class="n">count</span> <span class="o">+</span> <span class="mi">1</span>  
<span class="k">print</span> <span class="s">&quot;Total number of reads is: </span><span class="si">%i</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">))</span>  
<span class="k">print</span> <span class="s">&quot;Filtered reads is: </span><span class="si">%i</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">count</span><span class="p">)</span></code></pre></div>

<p>Although I&#39;ve learned a lot there were a problem that delayed the class progression: despite the organizers warning that the course was only for people that already knew Python basics, most of the class had not worked with python and a few had never programmed or even used command line to create a directory. This is not the organizer&#39;s or the instructor&#39;s fault and only delayed us in the first day. Actually those of us who already knew a bit of programming carried on with the exercises on our own.</p>

<p><strong>BioPython</strong></p>

<p>First let get out of the way that after the course ended, and for my own purposes - mostly analysis of NGS data - I did not see what could be done with BioPython that I could not already do with a combination of R and shell scripting. That said, parsing of files, for instance FASTA and even SAM appears to be faster and more intuitive than using R/Bioconductor packages. Parsing of files in Python appears to be much much simpler than in R. Of course there is a reason for that: R is for statistical analysis and was not developed to deal with strings. On the other hand Python has plenty of neat built-in functions to deal with strings that makes it much more powerful to deal with DNA/RNA/Protein sequences. BioPhyton seems to add even greater power for this.</p>

<p>Another point in favour of BioPhyton is its documentation which appears to be of a better standard than that of the average Bioconductor package. I&#39;ve also quite liked the possibility of wrapping any variable in help() to check with function can be used with that particular variable - simple and affective.</p>

<p>Plotting is ok in Python with Matplotlib but its aesthetics are not even close to those of ggplot in R.</p>

<p><strong>NGS data in BioPython</strong></p>

<p>Before doing the course I did some search on what was available for the analysis of NGS data using Python and got the impression that Python programmers appeared to have disregarded this technology. There were plenty of scripts in Awk, Perl, etc to parse, manipulate and analyse SAM/BAM/FastQ files, but not in Python. I was wrong. During the course I&#39;ve found out that BioPhtyton has <a href="http://www-huber.embl.de/users/anders/HTSeq/doc/index.html" title="HTSeq">HTSeq</a>. I have yet to explore it completely bu tit looks like it could be a nice tool for quick QC and count reads over features - something I am yet to master in R.</p>

<p>All in all I think I will be using Python a lot more  in the future but still doing most of analysis with R and shell scripting.</p>

  </article>

</div>

<div id="disqus_thread"></div>
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES * * */
    var disqus_shortname = 'adomingues-blog-io';

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">Moving To The Dark Side</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col  footer-col-1">
        <ul class="contact-list">
          <li>Moving To The Dark Side</li>
          <li><a href="mailto:amjdomingues@gmail.com">amjdomingues@gmail.com</a></li>
        </ul>
      </div>

      <div class="footer-col  footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/adomingues">
              <span class="icon  icon--github">
                <svg viewBox="0 0 16 16">
                  <path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/>
                </svg>
              </span>

              <span class="username">adomingues</span>
            </a>
          </li>
          

          
        </ul>
      </div>

      <div class="footer-col  footer-col-3">
        <p class="text"></p>
      </div>
    </div>

  </div>

</footer>


  </body>

  <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-60277379-1', 'auto');
  ga('send', 'pageview');

</script>


</html>
